### PR TITLE
Multi-hostname support, fixes #620

### LINKED ISSUE #620: Allow multiple host name aliases for one project
## What happened (or feature request):

* Feature request

## Motivation

It would be nice to be able to specify multiple additional (local) hostnames in configuration
which should be used for routing to the correct dev instance and used for the web server configuration (in case the web server does not accept arbitrary hostnames)

That is useful for applications that can deal with multiple host names in one instance (like TYPO3). It is a fairly common occurrence in TYPO3. For Drupal developers, this is akin to a domain access approach (one codebase, one database, multiple hostnames). Additionally, subdomains can be used for a multi-lingual approach.

## Acceptance Criteria

* End-users can add a hosts: entry in .ddev/config.yml
* Each entry within .ddev/config.yml gets added to /etc/hosts
* Each entry within .ddev/config.yml is used when reporting the site from ddev list, describe, and start.
* Each entry gets a local self-signed cert and appropriate nginx configs.
* ddev-ui and the json output it reads will need an array output.
* Each entry uses the same HTTP/HTTPS ports (that is we're not allowing per domain port configs)

## Related Issues

* Subdomain, Local DNS support: Don't require hosts file manipulation #416

### PR DESCRIPTION
**Pull request by rfay**
## The Problem/Issue/Bug:

We need to support multiple hostnames for multisite applications for #620 

## How this PR Solves The Problem:

Unlimited hostnames.

From the config.yaml self-documentation:
```
#additional_hostnames:
# - somename
# - someothername
# would provide http and https URLs for "somename.ddev.local"
# and "someothername.ddev.local".
```

## Manual Testing Instructions:

* Delete your .ddev/docker-compose.yaml
* Optionally run `ddev config` for the extra documentation it will put in config.yaml
* Add some additional_hostnames to your project, use `ddev start` and `ddev describe` to explore and hit them.

## Related Issue Link(s):

OP #620

### CODE REVIEW

**Comment by rfay** [comment]
This is actually ready to be taken for a test spin already. Still needs tests and some docs.

**Comment by rickmanelius** [comment]
Perfect! Will address today.

File: pkg/ddevapp/templates.go [hunk 1]
```diff
@@ -88,15 +88,46 @@ networks:
...
+
+# webimage: <docker_image>  # nginx/php docker image. 
+# dbimage: <docker_image>  # mariadb docker image. 
+# dbaimage: <docker_image>
+
+# router_http_port: <port>  # Port to be used for http (defaults to port 80)
+# router_https_port: <port> # Port for https (defaults to 443)
+
+# additional_hostnames[]  # For example, 
+#   additional_hostnames["name1.ddev.local", "othername.some.other.domain"]
```
[the remark this answers is not in the document]
Comment (rfay) [reply pkg/ddevapp/templates.go]: No, this shouldn't need a ":", but perhaps we shouldn't even expose this format, and just use the list format.

File: pkg/ddevapp/templates.go [hunk 2]
```diff
@@ -88,15 +88,46 @@ networks:
       name: ddev_default
 `
 
-// HookTemplate is used to add example hooks usage
-const HookTemplate = `
-# Certain ddev commands can be extended to run tasks before or after the ddev
-# command is executed.
+// ConfigInstructions is used to add example hooks usage
```
[the remark this answers is not in the document]
Comment (rfay) [reply pkg/ddevapp/templates.go]: Please give more of your suggestion here. "might be good to split" is a little hard to parse :)

File: pkg/ddevapp/templates.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rfay) [reply pkg/ddevapp/templates.go]: Yup, you're right.

**Comment by rfay** [comment]
Thanks for taking it for a spin Rick!

> Adding in a real, resolvable TLD doesn't seem to update the host machines DNS cache

This isn't my experience. If I use `additional_hostnames: ["randyfay.com"]` it adds randyfay.com with ip 127.0.0.1. Please give an example to replicate this problem.

> We still show the base project hostnames. Not an issue

We'd want to right? Trying not to change existing behavior?

> Collisions. I know we did some level of testing on making sure two sites were not up at the same time. I'm not sure that this is something we can or can't do

I think this is a potential issue, and am sure that our current collision avoidance (which is broken, IIRC) would not catch this at all.

> SSL. Seems to work, but again would be an issue if someone used a public TLD I'd imagine

This works fine, as expected. Using a real TLD is going to cause our users lots of heartache when they forget it's in their hosts file though.

randyfay com randy fay s drupal blog 2018-03-29 08-50-55

> Wildcard. I haven't tested, but curious if you think this would be in bounds of this request?

I don't think I understand your question. But if it's about having a *real* wildcard, we don't have any way to do that while we're tied to /etc/hosts because /etc/hosts has no concept of wildcard. However, people can add as many hostnames as they want. So www1.rickmanelius.local and www2.rickmanelius.com and www3.rickmanelius.somewhere can all be added.

**Comment by rickmanelius** [comment]
Hi @rfay. Responding in parts.

### DNS

I attempted to reconfigure a Drupal 7 site by updating config.yml and I now see "127.0.0.1 localhost randyfay.com" in /etc/hosts on my machine. However, running nslookup shows the following

```
➜  drupal-7.57 nslookup randyfay.com
Server:		192.168.1.1
Address:	192.168.1.1#53

Non-authoritative answer:
Name:	randyfay.com
Address: 173.255.248.186
```

And when I attempt to visit it I go to your site. I've attempted to flush cache (https://help.dreamhost.com/hc/en-us/articles/214981288-Flushing-your-DNS-cache-in-Mac-OS-X-and-Linux) and I've also tried across all 3 browsers, same outcome when going to http://randyfay.com.

### Smaller items:

* Default behavior: Agreed that we should keep as is for now.
* Collisions: Realistically, this in only a problem if someone attempts to bring up two versions of one codebase. It's highly unlikely that two different codebases would point to the same domain.
* SSL. I used a different project.ddev.local and verified that I can get through with SSL, so that's not an issue.
* Wildcard. This might be more appropriate to address #416 when we have public DNS support. I was just more curious if there was any incompatibilities.

### Summary

I think to unblock this for me I simply need to figure out why using an actual domain like randyfay.com doesn't properly route me to 127.0.01. The other items seem like they are either working or workable and shouldn't block this.

**Comment by siliconmeadow** [comment]
I gave this a spin on my Arch Linux laptop and it seemed to work ok. I then copied the version I'd compiled on the laptop to a CoreOS instance on GCP compute and it worked fine there too.

**Comment by rfay** [comment]
Thanks so much for the feedback @siliconmeadow! It may not be in its final form, but your feedback sure helps!

**Comment by siliconmeadow** [comment]
Regarding @rickmanelius 's issues with FQDN weirdness, I found similar, yet when putting in http://siliconmeadow.net, I used brought up the local ddev site:

```
Your project can be reached at http://sm.ddev.local, https://sm.ddev.local, http://siliconmeadow.net, https://siliconmeadow.net 
rsheppard@damnyank ~/projects/sm_d8/site % nslookup siliconmeadow.net     
Server:		192.168.1.254
Address:	192.168.1.254#53

Non-authoritative answer:
Name:	siliconmeadow.net
Address: 89.234.24.35

rsheppard@damnyank ~/projects/sm_d8/site %
```
hosts file entry:
`127.0.0.1 damnyank localhost.localdomain localhost sm.ddev.local flobble.local siliconmeadow.net`

### Further weirdness:

I shutdown the ddev site and reconfigured it by replacing `siliconmeadow.net` with `randyfay.com` in the config file and started it again.
`127.0.0.1 damnyank localhost.localdomain localhost sm.ddev.local flobble.local siliconmeadow.net randyfay.com`

When I pointed my browser at http://randyfay.com, it forced me to https, but then displayed the local https version of the site.

**Comment by rfay** [comment]
The issues pointed out in name resolution by @rickmanelius and @siliconmeadow really have to do only with unpredictable name resolution, nothing that ddev can control in any way. It's a good argument for allowing only *.ddev.local domain names, which is certainly on the table. However it would prevent people from the occasional odd time that they *have* to or *think* they have to use a real name. 

I see two paths here:
1. Offer two ways of supporting random hostnames, one that is the bare hostname and does "hostname.ddev.local", and the other (discouraged) that allows use of the full domain name.
2. Just stop allowing the full domain name and allow hostname only. It would still be possible (I think) to do the full domain name using docker-compose.yaml techniques, we could either document that or not.

**Comment by rickmanelius** [comment]
As suggested by @rfay, I wouldn't be opposed to using option #2 (restrict to hostnames) until we address #416, where we can open back up to FQDN. Even then we can't 100% fix the unpredictable behavior based on DNS caching, so we'd have to think of some way to message the end-user about that possibility as well as a link to an FAQ to prod it along (like I was able to by modifying /etc/host by hand and creating its own line).

TL;DR Let's go with #2 for now to open up the feature and we can push for FQDN later once we see how many users even use this and if they are pushing for more.

**Comment by rickmanelius** [comment]
Still needs a code/test review from @cweagans after we remove the FQDN.

**Comment by cweagans** [comment]
One minor nit: there are a few places here where multi-word strings in a YAML file are not quoted. I _think_ the spec recommends it, but most parsers handle unquoted strings gracefully, so not sure it's worth worrying about.
